File: solutions/python/rna-strand-folding-score.py
"""
RNA Strand Folding Score
Difficulty: Medium
Topics: dynamic-programming, string, biology

In RNA biology, nucleotide sequences can fold and form base pairs. Given a string 
representing an RNA sequence containing only characters 'A', 'U', 'G', and 'C', 
calculate the maximum folding score. A folding score is calculated by the number 
of valid base pairs that can be formed under these rules:

1. A can pair with U
2. G can pair with C  
3. Each nucleotide can pair with at most one other nucleotide
4. Base pairs cannot cross (if i-j and k-l are pairs, we cannot have i < k < j < l)

Return the maximum possible folding score.

Examples:
- AUUGC → 2 (A-U at (0,1) and G-C at (3,4))
- GCAU → 1 (either G-C or A-U, not both due to crossing)
- AAUU → 2 (both A's pair with U's without crossing)
"""

def maxRNAFoldingScore(sequence: str) -> int:
    """
    Calculate the maximum RNA folding score for a given sequence.
    
    Args:
        sequence (str): RNA sequence containing only 'A', 'U', 'G', 'C'
    
    Returns:
        int: Maximum possible folding score
    """
    n = len(sequence)
    
    # Initialize DP table with zeros
    # dp[i][j] represents max score for substring from i to j
    dp = [[0] * n for _ in range(n)]
    
    # Helper function to check if two bases can pair
    def can_pair(x: str, y: str) -> bool:
        return (x == 'A' and y == 'U') or \
               (x == 'U' and y == 'A') or \
               (x == 'G' and y == 'C') or \
               (x == 'C' and y == 'G')
    
    # Fill DP table
    # l is the length of substring being considered
    for l in range(2, n + 1):
        # i is the start position of substring
        for i in range(n - l + 1):
            j = i + l - 1  # j is end position of substring
            
            # Case 1: Don't pair position i, move to i+1
            dp[i][j] = dp[i + 1][j]
            
            # Case 2: Pair i with some position k where i < k <= j
            for k in range(i + 1, j + 1):
                if can_pair(sequence[i], sequence[k]):
                    score = 1  # Score for pairing i with k
                    if k < j:
                        score += dp[k + 1][j]  # Add score for substring after k
                    if i + 1 < k:
                        score += dp[i + 1][k - 1]  # Add score for substring between i and k
                    dp[i][j] = max(dp[i][j], score)
    
    return dp[0][n - 1]


# Test cases
def test_maxRNAFoldingScore():
    """Test the RNA folding score function"""
    
    # Test case 1: Basic example
    assert maxRNAFoldingScore("AUUGC") == 2
    
    # Test case 2: Non-crossing pairs can coexist  
    assert maxRNAFoldingScore("GCAU") == 2  # G-C(0,1) and A-U(2,3) don't cross
    
    # Test case 3: Multiple A-U pairs (corrected)
    assert maxRNAFoldingScore("AAUU") == 2  # A(0)-U(3) and A(1)-U(2) don't cross
    
    # Test case 4: Complex case (fixed)
    assert maxRNAFoldingScore("AUGCAU") == 3  # A-U, G-C, A-U
    
    # Test case 5: Maximum pairing
    assert maxRNAFoldingScore("GGGGCCCC") == 4
    
    # Test case 6: Example with limited pairing options  
    assert maxRNAFoldingScore("AUGU") == 1  # Only A-U pairing possible
    
    # Test case 7: No pairs possible
    assert maxRNAFoldingScore("AC") == 0
    
    # Test case 8: Complex optimal case (fixed)
    assert maxRNAFoldingScore("GCAUGCAU") == 4  # Multiple optimal pairings
    
    print("All tests passed!")


if __name__ == "__main__":
    test_maxRNAFoldingScore()
    
    # Example usage
    examples = ["AUUGC", "GCAU", "AAUU", "GCAUGCAU"]
    for seq in examples:
        score = maxRNAFoldingScore(seq)
        print(f"Sequence: {seq}, Max Folding Score: {score}")

File: solutions/react-ts/rna-strand-folding-score.ts
/**
 * Problem: RNA Strand Folding Score
 * Difficulty: Medium
 * Topics: dynamic-programming, string, biology
 * 
 * Description:
 * In RNA biology, nucleotide sequences can fold and form base pairs. Given a string representing an RNA sequence containing only characters 'A', 'U', 'G', and 'C', calculate the maximum folding score. A folding score is calculated by the number of valid base pairs that can be formed under these rules:

1. A can pair with U
2. G can pair with C
3. Each nucleotide can pair with at most one other nucleotide
4. Base pairs cannot cross (if i-j and k-l are pairs, we cannot have i < k < j < l)

Return the maximum possible folding score.
 * 
 * Examples:
 * TODO: Add examples
 * 
 * Constraints:
 * TODO: Add constraints
 */

interface TestCase {
  input: any;
  expectedOutput: any;
  description: string;
}

class Solution {
  /**
   * Main solution method
   * 
   * Time Complexity: O(?)
   * Space Complexity: O(?)
   * 
   * @param args Problem-specific arguments
   * @returns Problem-specific return value
   */
  solve(...args: any[]): any {
    // Implementation here
    return null;
  }
}

// Test cases
const testCases: TestCase[] = [
  {
    input: [], // Replace with actual input
    expectedOutput: null, // Replace with expected output
    description: "Example 1"
  },
  {
    input: [], // Replace with actual input
    expectedOutput: null, // Replace with expected output
    description: "Example 2"
  }
];

// Test runner
function runTests(): void {
  const solution = new Solution();
  let passed = 0;
  let total = testCases.length;

  console.log(`Running ${total} test cases...\n`);

  testCases.forEach((testCase, index) => {
    try {
      const result = solution.solve(...testCase.input);
      const isEqual = JSON.stringify(result) === JSON.stringify(testCase.expectedOutput);
      
      if (isEqual) {
        console.log(`✅ Test ${index + 1}: ${testCase.description} - PASSED`);
        passed++;
      } else {
        console.log(`❌ Test ${index + 1}: ${testCase.description} - FAILED`);
        console.log(`   Expected: ${JSON.stringify(testCase.expectedOutput)}`);
        console.log(`   Got: ${JSON.stringify(result)}`);
      }
    } catch (error) {
      console.log(`❌ Test ${index + 1}: ${testCase.description} - ERROR`);
      console.log(`   Error: ${error}`);
    }
  });

  console.log(`\n${passed}/${total} tests passed`);
}

// Export for testing framework
export { Solution, testCases, runTests };

// Run tests if this file is executed directly
if (require.main === module) {
  runTests();
}